src/build_helper/src/compiletest.rs

```diff
@@ -0,0 +1,130 @@
+//! Types representing arguments to compiletest.
+
+use std::fmt;
+use std::str::FromStr;
+
+pub use self::Mode::*;
+
+macro_rules! string_enum {
+    ($(#[$meta:meta])* $vis:vis enum $name:ident { $($variant:ident => $repr:expr,)* }) => {
+        $(#[$meta])*
+        $vis enum $name {
+            $($variant,)*
+        }
+
+        impl $name {
+            $vis const VARIANTS: &'static [Self] = &[$(Self::$variant,)*];
+            $vis const STR_VARIANTS: &'static [&'static str] = &[$(Self::$variant.to_str(),)*];
+
+            $vis const fn to_str(&self) -> &'static str {
+                match self {
+                    $(Self::$variant => $repr,)*
+                }
+            }
+        }
+
+        impl fmt::Display for $name {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                fmt::Display::fmt(self.to_str(), f)
+            }
+        }
+
+        impl FromStr for $name {
+            type Err = String;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                match s {
+                    $($repr => Ok(Self::$variant),)*
+                    _ => Err(format!(concat!("unknown `", stringify!($name), "` variant: `{}`"), s)),
+                }
+            }
+        }
+    }
+}
+
+// Make the macro visible outside of this module, for tests.
+#[cfg(test)]
+pub(crate) use string_enum;
+
+string_enum! {
+    #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+    pub enum Mode {
+        Pretty => "pretty",
+        DebugInfo => "debuginfo",
+        Codegen => "codegen",
+        Rustdoc => "rustdoc",
+        RustdocJson => "rustdoc-json",
+        CodegenUnits => "codegen-units",
+        Incremental => "incremental",
+        RunMake => "run-make",
+        Ui => "ui",
+        RustdocJs => "rustdoc-js",
+        MirOpt => "mir-opt",
+        Assembly => "assembly",
+        CoverageMap => "coverage-map",
+        CoverageRun => "coverage-run",
+        Crashes => "crashes",
+    }
+}
+
+impl Default for Mode {
+    fn default() -> Self {
+        Mode::Ui
+    }
+}
+
+impl Mode {
+    pub fn aux_dir_disambiguator(self) -> &'static str {
```

Comment:
I don't feel like these two helpers belong in build_helpers?

Reply:
that would require turning them into free functions.  do you want me to do that?

Reply:
Yes, please do move them as free functions into compiletest, this is an impl detail that doesn't feel like they belong here